File: evaluasiStack.py
def precedence(op):
    if op in ['+', '-']:
        return 1
    if op in ['*', '/']:
        return 2
    return 0


def infix_to_postfix(expression):
    stack = []
    postfix = []
    number = ""

    print("\nProses Push-Pop Stack (Infix -> Postfix):")

    for c in expression:
        # BUG FIX 1: Abaikan spasi agar ekspresi seperti "3 + 4" tetap valid
        if c == ' ':
            continue

        if c.isdigit() or c == '.':  # BUG FIX 2: Tambah support angka desimal
            number += c

        else:
            if number:
                postfix.append(number)
                print("Output:", " ".join(postfix))
                number = ""

            if c == '(':
                stack.append(c)
                print("Push:", c)

            elif c == ')':
                while stack and stack[-1] != '(':
                    postfix.append(stack.pop())
                    print("Pop -> Output:", " ".join(postfix))
                stack.pop()  # Buang '('

            elif c in "+-*/":
                while stack and stack[-1] != '(' and precedence(stack[-1]) >= precedence(c):
                    postfix.append(stack.pop())
                    print("Pop -> Output:", " ".join(postfix))

                stack.append(c)
                print("Push:", c)

    if number:
        postfix.append(number)
        print("Output:", " ".join(postfix))

    while stack:
        postfix.append(stack.pop())
        print("Pop -> Output:", " ".join(postfix))

    return postfix


def evaluate_postfix(postfix):
    stack = []

    print("\nStep Evaluasi Postfix:")

    for token in postfix:
        # BUG FIX 3: Ganti token.isdigit() -> try/except float()
        # isdigit() gagal untuk angka desimal ("3.5") dan angka negatif ("-3")
        try:
            stack.append(float(token))
            print("Push:", token)

        except ValueError:
            b = stack.pop()
            a = stack.pop()

            if token == '+':
                result = a + b
            elif token == '-':
                result = a - b
            elif token == '*':
                result = a * b
            elif token == '/':
                result = a / b

            stack.append(result)
            print("Pop", a, "dan", b, "-> Push hasil:", result)

    final = stack.pop()
    # BUG FIX 4: Tampilkan int jika hasilnya bilangan bulat (misal 6.0 -> 6)
    return int(final) if final == int(final) else final


# MAIN
infix = input("Masukkan ekspresi infix: ")

postfix = infix_to_postfix(infix)

print("\nPostfix Expression:", " ".join(postfix))

hasil = evaluate_postfix(postfix)

print("\nHasil Akhir:", hasil)

File: EvaluasiStack.java
import java.util.Stack;
import java.util.Scanner;

public class EvaluasiStack {

    static int precedence(char op) {
        if (op == '+' || op == '-') return 1;
        if (op == '*' || op == '/') return 2;
        return 0;
    }

    static String infixToPostfix(String exp) {
        Stack<Character> stack = new Stack<>();
        // BUG FIX 4: Pakai StringBuilder, bukan String += (lebih efisien)
        StringBuilder postfix = new StringBuilder();

        System.out.println("\nProses Push-Pop Stack (Konversi Infix -> Postfix):");

        for (int i = 0; i < exp.length(); i++) {
            char c = exp.charAt(i);

            // BUG FIX 2: Skip spasi agar "3 + 4" tetap valid
            if (c == ' ') continue;

            // BUG FIX 1: Tangani angka multi-digit (misal: 12, 100)
            // Baca semua digit berurutan, gabung jadi satu token dengan separator spasi
            if (Character.isDigit(c)) {
                StringBuilder number = new StringBuilder();
                while (i < exp.length() && Character.isDigit(exp.charAt(i))) {
                    number.append(exp.charAt(i));
                    i++;
                }
                i--; // Kembalikan index karena loop for akan i++ lagi

                postfix.append(number).append(" ");
                System.out.println("Output: " + postfix.toString().trim());
            }
            else if (c == '(') {
                stack.push(c);
                System.out.println("Push: " + c);
            }
            else if (c == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    postfix.append(stack.pop()).append(" ");
                    System.out.println("Pop -> Output: " + postfix.toString().trim());
                }
                stack.pop(); // Buang '('
            }
            else if (c == '+' || c == '-' || c == '*' || c == '/') {
                // BUG FIX 3: Tambah guard stack.peek() != '(' agar '(' tidak di-pop
                // saat membandingkan precedence
                while (!stack.isEmpty() && stack.peek() != '(' && precedence(stack.peek()) >= precedence(c)) {
                    postfix.append(stack.pop()).append(" ");
                    System.out.println("Pop -> Output: " + postfix.toString().trim());
                }
                stack.push(c);
                System.out.println("Push: " + c);
            }
        }

        while (!stack.isEmpty()) {
            postfix.append(stack.pop()).append(" ");
            System.out.println("Pop -> Output: " + postfix.toString().trim());
        }

        return postfix.toString().trim();
    }

    // BUG FIX 1 (lanjutan): evaluatePostfix sekarang baca token per token (split spasi)
    // karena postfix sudah berformat "12 3 +" bukan "123+"
    static double evaluatePostfix(String postfix) {
        Stack<Double> stack = new Stack<>();

        System.out.println("\nStep Evaluasi Postfix:");

        String[] tokens = postfix.split(" ");

        for (String token : tokens) {
            if (token.isEmpty()) continue;

            // Cek apakah token adalah angka (bukan operator)
            try {
                double num = Double.parseDouble(token);
                stack.push(num);
                System.out.println("Push: " + token);
            } catch (NumberFormatException e) {
                double b = stack.pop();
                double a = stack.pop();
                double result = 0;

                switch (token.charAt(0)) {
                    case '+': result = a + b; break;
                    case '-': result = a - b; break;
                    case '*': result = a * b; break;
                    case '/':
                        // BUG FIX 5: Tambah pengecekan division by zero
                        if (b == 0) throw new ArithmeticException("Error: Pembagian dengan nol!");
                        result = a / b;
                        break;
                }

                stack.push(result);
                System.out.println("Pop " + a + " dan " + b + " -> Push hasil: " + result);
            }
        }

        return stack.pop();
    }

    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        System.out.print("Masukkan ekspresi infix: ");
        String infix = input.nextLine();

        String postfix = infixToPostfix(infix);

        System.out.println("\nPostfix Expression: " + postfix);

        double hasil = evaluatePostfix(postfix);

        // Tampilkan sebagai int jika hasilnya bilangan bulat (misal 6.0 -> 6)
        if (hasil == (long) hasil) {
            System.out.println("\nHasil Akhir: " + (long) hasil);
        } else {
            System.out.println("\nHasil Akhir: " + hasil);
        }

        input.close();
    }
}